core/templates/fixed_vector.h

```diff
@@ -0,0 +1,157 @@
+/*                         This file is part of:                          */
+/*                             GODOT ENGINE                               */
+/*                        https://godotengine.org                         */
+/**************************************************************************/
+/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#pragma once
+
+// A high speed Vector of fixed capacity.
+// Especially useful if you need to create an array on the stack, to
+// prevent dynamic allocations (especially in bottleneck code).
+
+// Choose CAPACITY such that it is enough for all elements that could be added through all branches.
+
+// Set force_trivial to true to gain a slight performance improvement on resize calls
+//  if the default value of the type can be ignored.
+template <class T, uint32_t CAPACITY, bool force_trivial = false>
+class FixedVector {
+	// This declaration allows us to access other FixedVector's private members.
+	template <class T_, uint32_t CAPACITY_, bool force_trivial_>
+	friend class FixedVector;
+
+	static_assert(!force_trivial || std::is_trivially_destructible_v<T>, "force_trivial may only be true if the type is trivially destructible.");
+
+	uint32_t _size = 0;
+	alignas(T) uint8_t _data[CAPACITY * sizeof(T)];
```

Comment:
if you wrap this inside a struct and declare one like : 
```
struct FixedVectorInner
{
     uint32_t _size = 0;
     alignas(T) uint8_t _data[CAPACITY * sizeof(T)];
};
FixedVectorInner inner;
```
Then you don't need the DATA_PADDING information as you can just memcpy with adress of inner and sizeof(inner).  Just a remark.

Reply:
Well you still can with `memcpySize = &(inner.data + size * sizeof(T)) - &inner` but that's not really better than the current one, agreed.

Reply:
That's also an interesting idea! C++ sure can do the same thing in so many different ways 😄
I think for now I still like the data padding best, because it's only awkward in its construction but looks pretty normal elsewhere.